src/sources/pipe.rs

```diff
@@ -0,0 +1,247 @@
+    serde::default_decoding,
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+/// Configuration for the `pipe` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PipeConfig {
+    /// The maximum buffer size, in bytes, of incoming messages.
+    ///
+    /// Messages larger than this are truncated.
+    #[serde(default = "crate::serde::default_max_length")]
+    pub max_length: usize,
+
+    /// Overrides the name of the log field used to add the current hostname to each event.
+    ///
+    /// The value will be the current hostname for wherever Vector is running.
+    ///
+    /// By default, the [global `host_key` option](https://vector.dev/docs/reference/configuration//global-options#log_schema.host_key) is used.
+    pub host_key: Option<String>,
+
+    #[configurable(derived)]
+    pub framing: Option<FramingConfig>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    pub decoding: DeserializerConfig,
+
+    /// The file descriptor number to read from.
+    pub fd: c_int,
+}
+
+inventory::submit! {
+    SourceDescription::new::<PipeConfig>("pipe")
+}
+
+impl GenerateConfig for PipeConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(indoc! {r#"
+            fd = 500
+        "#})
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "pipe")]
+impl SourceConfig for PipeConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        pipe_source(self.clone(), cx.shutdown, cx.out)
+    }
+
+    fn outputs(&self) -> Vec<Output> {
+        vec![Output::default(self.decoding.output_type())]
+    }
+
+    fn source_type(&self) -> &'static str {
+        "pipe"
+    }
+
+    fn resources(&self) -> Vec<Resource> {
+        vec![Resource::Fd(0)]
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+pub fn pipe_source(
+    config: PipeConfig,
+    shutdown: ShutdownSignal,
+    mut out: SourceSender,
+) -> crate::Result<super::Source> {
+    let mut pipe = io::BufReader::new(unsafe { File::from_raw_fd(config.fd) });
+    let host_key = config
+        .host_key
+        .unwrap_or_else(|| log_schema().host_key().to_string());
+    let hostname = crate::get_hostname().ok();
+
+    let framing = config
+        .framing
+        .unwrap_or_else(|| config.decoding.default_stream_framing());
+    let decoder = DecodingConfig::new(framing, config.decoding).build();
+
+    let (mut sender, receiver) = mpsc::channel(1024);
+
+    // Spawn background thread with blocking I/O to process fd.
+    //
+    // This is recommended by Tokio, as otherwise the process will not shut down
+    // until another newline is entered. See
+    // https://github.com/tokio-rs/tokio/blob/a73428252b08bf1436f12e76287acbc4600ca0e5/tokio/src/io/stdin.rs#L33-L42
+    thread::spawn(move || {
+        info!("Capturing fd.");
```

Comment:
Might be useful to see the fd number here:

```suggestion
        info!("Capturing fd {}.", self.fd);
```